File: binary_search/binary_search.js
function enumerate(elements) {
    let result = [];
    elements.forEach(function(item, i) {
        result.push([i, item]);
    });
    return result;
}


function binarySearch(collection, target) {

    function inner(collection, target) {
        if (collection.length === 0) {
            return false;
        }
        let middleIndex = parseInt(collection.length/2);
        let candidate = collection[middleIndex];
        if (candidate[1] === target) {
            return candidate[0];
        }
        if (candidate[1] > target) {
            return inner(collection.slice(0, middleIndex), target);
        }
        if (candidate[1] < target) {
            return inner(collection.slice(middleIndex+1), target);
        }
    }

    return inner(enumerate(collection), target);
}


// TEST
let myList = [...Array(10000).keys()];
console.log(binarySearch(myList, 5555));

File: algorithms/binary_search.py
def binary_search(collection, target):

    def inner(collection, target):
        if not collection:
            return False
        middle_index = len(collection)//2
        candidate = collection[middle_index]
        if candidate[1] == target:
            return candidate[0]
        if candidate[1] > target:
            return inner(collection[:middle_index], target)
        if candidate[1] < target: 
            return inner(collection[middle_index+1:], target)

    return inner(list(enumerate(collection)), target)


# TEST
my_list = list(range(10000))
assert binary_search(my_list, 5555) == 5555
